### PR TITLE
Add `filter_map`, `dedup` and some variants to the stdlib

### LINKED ISSUE #1958: Some utilities in `std.array`:  `remove_duplicates` and `filter_map`
**Is your feature request related to a problem? Please describe.**

I need to take values that fit in to nice high-level schema and re-organize them for ad-hoc "back ends" like Ansible inventories. Some more array processing functions would help in this transformation.

**Describe the solution you'd like**

* A function `std.array.remove_duplicates`. This could be written to take an equality function and likely have the type `forall a (a -> a -> Bool) -> Array a-> Array a`. It could also use built-in equality testing for a simpler definition like `forall a Array a -> Array a`.
* A function `std.array.filter_map`. I'm not sure if the type of function is express-able -- i.e. it would need a type _like_ `forall a b. (a -> b) -> Array a -> Array b` but `b` could be a `Bool` (with the only acceptable value being `false`).

**Describe alternatives you've considered**

I have workarounds that bloat the code a bit.

**Additional context**

Hopefully these are fairly understood functions -- they definitely exist in other functional languages. The type system of Nickel is a bit different than what I'm used to so perhaps the types would need adjustment.

### PR DESCRIPTION
**Pull request by yannham**
Closes #1958. Adds `filter_map`, `dedup`, as well as two variants of `dedup` that avoid the quadratic behavior of the purely equality-based vanilla version, `sort_dedup` and `hash_dedup`.

### CODE REVIEW

**Comment by yannham** [comment]
It seems we now blow the stack on Windows (1MB), and it's can't be really tweaked from tests - I tried to wrap all main tests as a scoped thread with larger stack but it didn't work so I suspect this is the separate LSP command. This is annoying and also probably means the standalone LSP binary is also overflowing the stack on Windows.

File: core/stdlib/std.ncl [hunk 1]
```diff
@@ -917,6 +917,181 @@
...
+
+        # Hash function
+
+        `hash_dedup` uses a "hash" function mapping array elements to strings.
+        This function is used to efficiently check if an element has already
+        been seen before (linear time amortized in the size of array versus
+        quadratic for `std.array.dedup`).
+
+        The hash function must separate distinct elements with very high
+        probability. If two elements have the same hash, they will be considered
```
[the remark this answers is not in the document]
Comment (yannham) [reply core/stdlib/std.ncl:1055]: You're right, it's a logic error to not separate unequal keys. In some sense the hash function **is** the notion of equality at hand. Maybe I shouldn't call this a hash function at all?

File: core/stdlib/std.ncl [hunk 2]
```diff
@@ -917,6 +917,181 @@
...
+          (fun acc x =>
+            if elem x acc then
+              acc
+            else
+              acc @ [x]
+          )
+          []
+          array,
+
+    sort_dedup
```
[the remark this answers is not in the document]
Comment (yannham) [reply core/stdlib/std.ncl:988]: Sure, we can factor out `dedup_sorted`. I would still keep `sort_dedup` as a handy alias, though.

**Comment by yannham** [comment]
After discussing with @jneem through other channels, we agreed that `hash_dedup` has a strange interface and will be split off this pull request. It should make a come back but without having to provide an explicit hash function; either using a built-in that we'll add to Nickel, or as a middle ground waiting for that, use `std.serialize` automatically.

**Review by jneem — APPROVED** [review]

**Pull request merged**